python/ray/includes/gcs_client.pxi

```diff
@@ -360,3 +488,151 @@ cdef class NewGcsClient:
+        node_table_data = {}
+        for c_proto in c_data:
+            b = c_proto.SerializeAsString()
+            proto = gcs_pb2.GcsNodeInfo()
+            proto.ParseFromString(b)
+            node_table_data[proto.node_id] = proto
+        return node_table_data, None
+    except Exception as e:
+        return None, e
+
+cdef convert_get_all_job_info(
+        CRayStatus status, c_vector[CJobTableData]&& c_data):
+    # -> Dict[JobID, gcs_pb2.JobTableData]
+    cdef c_string b
+    try:
+        check_status_timeout_as_rpc_error(status)
+        job_table_data = {}
+        for c_proto in c_data:
+            b = c_proto.SerializeAsString()
+            proto = gcs_pb2.JobTableData()
+            proto.ParseFromString(b)
+            job_table_data[proto.job_id] = proto
+        return job_table_data, None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_str_none_for_not_found(
+        CRayStatus status, const optional[c_string]& c_str):
+    # If status is NotFound, return None.
+    # If status is OK, return the value.
+    # Else, raise exception.
+    # -> Optional[bytes]
+    try:
+        if status.IsNotFound():
+            return None, None
+        check_status_timeout_as_rpc_error(status)
+        return c_str.value(), None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_multi_get(
+        CRayStatus status, const optional[unordered_map[c_string, c_string]]& c_map):
+    # -> Dict[str, str]
+    cdef unordered_map[c_string, c_string].const_iterator it
+    try:
+        check_status_timeout_as_rpc_error(status)
+        result = {}
+        it = c_map.value().const_begin()
+        while it != c_map.value().const_end():
+            key = dereference(it).first
+            value = dereference(it).second
+            result[key] = value
+            postincrement(it)
+        return result, None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_int(CRayStatus status, const optional[int]& c_int):
+    # -> int
+    try:
+        check_status_timeout_as_rpc_error(status)
+        return c_int.value(), None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_vector_str(
+        CRayStatus status, const optional[c_vector[c_string]]& c_vec):
+    # -> Dict[str, str]
+    try:
+        check_status_timeout_as_rpc_error(status)
+    except Exception as e:
+        return None, e
+    cdef const c_vector[c_string]* vec = &c_vec.value()
+    cdef c_vector[c_string].const_iterator it = dereference(vec).const_begin()
+    result = []
+    while it != dereference(vec).const_end():
+        result.append(dereference(it))
+        postincrement(it)
+    return result, None
+
+cdef convert_optional_bool(CRayStatus status, const optional[c_bool]& b):
+    # -> bool
+    try:
+        check_status_timeout_as_rpc_error(status)
+    except Exception as e:
+        return None, e
+    return b.value(), None
```

Comment:
I feel we should handle optional here. It's error-prone that this method is `convert_optional_bool` but it assumes it's never optional?

Reply:
```
Status InternalKVAccessor::Put(const std::string &ns,
                               const std::string &key,
                               const std::string &value,
                               bool overwrite,
                               const int64_t timeout_ms,
                               bool &added) {
  std::promise<Status> ret_promise;
  RAY_CHECK_OK(AsyncInternalKVPut(
      ns,
      key,
      value,
      overwrite,
      timeout_ms,
      [&ret_promise, &added](Status status, std::optional<int> added_num) {
        added = static_cast<bool>(added_num.value_or(0));
        ret_promise.set_value(status);
      }));
  return ret_promise.get_future().get();
}
```

For example, this handles the optional

Reply:
ok I will assert for each optional

Reply:
that optional check would never be used because its caller always populates:

```
  client_impl_->GetGcsRpcClient().InternalKVPut(
      req,
      [callback](const Status &status, const rpc::InternalKVPutReply &reply) {
        callback(status, reply.added_num());
      },
      timeout_ms);
```

but nonetheless I added assert for all optional converters